pkg/initializer/util/csrUtil.go

```diff
@@ -0,0 +1,163 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package util
+
+import (
+	"crypto/rsa"
+	"fmt"
+	certUtil "github.com/pingcap/tidb-operator/pkg/util"
+	certificates "k8s.io/api/certificates/v1beta1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/klog"
+	"time"
+)
+
+func CSRNameForService(serviceName, namespace string) string {
+	return fmt.Sprintf("%s.%s", serviceName, namespace)
+}
+
+func createApprovedCSR(serviceName, namespace string, ownerRefs []metav1.OwnerReference, kubeCli kubernetes.Interface, key *rsa.PrivateKey, timeout time.Duration) (*certificates.CertificateSigningRequest, error) {
+
+	csrName := CSRNameForService(serviceName, namespace)
+	err := waitUntilCsrDelete(csrName, kubeCli, timeout)
+	if err != nil {
+		return nil, err
+	}
+
+	klog.Infof("success to delete csr[%s]", csrName)
+	dnsNames := []string{
+		serviceName,
+		fmt.Sprintf("%s.%s", serviceName, namespace),
+		fmt.Sprintf("%s.%s.svc", serviceName, namespace),
+	}
+	requestPem, err := certUtil.GenerateCSRPem(key, serviceName, dnsNames)
+	if err != nil {
+		return nil, err
+	}
+
+	csr := generateCSR(serviceName, namespace, ownerRefs, requestPem)
+	createdCSR, err := kubeCli.CertificatesV1beta1().CertificateSigningRequests().Create(csr)
+	if err != nil {
+		return nil, err
+	}
+	klog.Infof("success to create csr[%s]", csrName)
+
+	_, err = approveCSR(createdCSR, kubeCli, timeout)
+	if err != nil {
+		return nil, err
+	}
+	err = waitUntilCsrApproved(csrName, kubeCli, timeout)
+	if err != nil {
+		return nil, err
+	}
+	klog.Infof("success to approve csr[%s]", csrName)
+
+	approvedCSR, err := kubeCli.CertificatesV1beta1().CertificateSigningRequests().Get(csrName, metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	return approvedCSR, nil
+}
+
+func approveCSR(csr *certificates.CertificateSigningRequest, kubeCli kubernetes.Interface, timeout time.Duration) (*certificates.CertificateSigningRequest, error) {
+	err := waitUntilCsrExist(csr.Name, kubeCli, timeout)
+	if err != nil {
+		return nil, err
+	}
+	csr = generateApproveCSR(csr)
+	approvedCSR, err := kubeCli.CertificatesV1beta1().CertificateSigningRequests().UpdateApproval(csr)
+	if err != nil {
+		return nil, err
+	}
+	return approvedCSR, nil
+}
+
+func generateCSR(service, namespace string, ownerRefs []metav1.OwnerReference, requestPem []byte) *certificates.CertificateSigningRequest {
+	csr := &certificates.CertificateSigningRequest{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            fmt.Sprintf("%s.%s", service, namespace),
```

Comment:
why not use the `CSRNameForService` func?

Reply:
or we can remove the func, which is better IMHO

Reply:
fix